### ISSUE TITLE
[BUG] Unable to compile cpp2 enum

### ISSUE THREAD

**Issue by MatthieuHernandez** [hsutter/cppfront#1304]
I have a compilation error with Clang 16.0.5 when I try to declare an enum in cpp2. I use cppfront compiled from main branch (ccf7011a4ef50eadb1266c5604bbd8d9ad0d87ea).

*compiler error*
```
clang  -IC:/Programming/Git/gopp2/../cppfront/include -O0 -g -Xclang -gcodeview -D_DEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrtd -std=gnu++20 -MD -MT CMakeFiles/gopp2.dir/generate/main.cpp.obj -MF CMakeFiles\gopp2.dir\generate\main.cpp.obj.d -o CMakeFiles/gopp2.dir/generate/main.cpp.obj -c C:/Programming/Git/gopp2/generate/main.cpp
In file included from ../src/main.cpp2:2:
In file included from ../src/engine.h2:2:
In file included from ../src/move.h2:1:
../src/color.h2:12:18: error: constexpr constructor never produces a constant expression [-Winvalid-constexpr]
constexpr Color::Color()
                 ^
../src/color.h2:13:51: note: read of non-constexpr variable 'Black' is not allowed in a constant expression
                                        : _value{ Black._value }{}
                                                  ^
../src/color.h2:7:36: note: declared here
inline CPP2_CONSTEXPR Color Color::Black{ 0 };
                                   ^
1 error generated.
```

*color.h2*
```cpp
Color: @enum<u8> type = {
    Black := 0;
    White := 1;
}
```

*color.h*
```cpp
#ifndef COLOR_H_CPP2
#define COLOR_H_CPP2

#define CPP2_IMPORT_STD          Yes

//=== Cpp2 type declarations ====================================================


#include "cpp2util.h"

#line 1 "../src/color.h2"
class Color;
#line 2 "../src/color.h2"
    

//=== Cpp2 type definitions and function declarations ===========================

#line 1 "../src/color.h2"
class Color {
private: cpp2::u8 _value; private: constexpr Color(cpp2::impl::in<cpp2::i64> _val);

private: constexpr auto operator=(cpp2::impl::in<cpp2::i64> _val) -> Color& ;
public: static const Color Black;
public: static const Color White;
public: [[nodiscard]] constexpr auto get_raw_value() const& -> cpp2::u8;
public: constexpr explicit Color();
public: constexpr Color(Color const& that);
public: constexpr auto operator=(Color const& that) -> Color& ;
public: constexpr Color(Color&& that) noexcept;
public: constexpr auto operator=(Color&& that) noexcept -> Color& ;
public: [[nodiscard]] auto operator<=>(Color const& that) const& -> std::strong_ordering = default;
public: [[nodiscard]] auto to_string_impl(cpp2::impl::in<std::string_view> prefix) const& -> std::string;
public: [[nodiscard]] auto to_string() const& -> std::string;
public: [[nodiscard]] auto to_code() const& -> std::string;
public: [[nodiscard]] static auto from_string(cpp2::impl::in<std::string_view> s) -> Color;
public: [[nodiscard]] static auto from_code(cpp2::impl::in<std::string_view> s) -> Color;

#line 4 "../src/color.h2"
};


//=== Cpp2 function definitions =================================================

#line 1 "../src/color.h2"

#line 1 "../src/color.h2"

constexpr Color::Color(cpp2::impl::in<cpp2::i64> _val)
                                                          : _value{ cpp2::unchecked_narrow<cpp2::u8>(_val) } {  }
constexpr auto Color::operator=(cpp2::impl::in<cpp2::i64> _val) -> Color&  { 
                                                          _value = cpp2::unchecked_narrow<cpp2::u8>(_val);
                                                          return *this; }
inline CPP2_CONSTEXPR Color Color::Black{ 0 };

inline CPP2_CONSTEXPR Color Color::White{ 1 };

[[nodiscard]] constexpr auto Color::get_raw_value() const& -> cpp2::u8 { return _value; }
constexpr Color::Color()
                                        : _value{ Black._value }{}
constexpr Color::Color(Color const& that)
                                              : _value{ that._value }{}
constexpr auto Color::operator=(Color const& that) -> Color& {
                                              _value = that._value;
                                              return *this;}
constexpr Color::Color(Color&& that) noexcept
                                              : _value{ std::move(that)._value }{}
constexpr auto Color::operator=(Color&& that) noexcept -> Color& {
                                              _value = std::move(that)._value;
                                              return *this;}
[[nodiscard]] auto Color::to_string_impl(cpp2::impl::in<std::string_view> prefix) const& -> std::string{

auto pref {cpp2::to_string(prefix)}; 
if ((*this) == Black) {return pref + "Black"; }
if ((*this) == White) {return cpp2::move(pref) + "White"; }
return "invalid Color value"; 
}

[[nodiscard]] auto Color::to_string() const& -> std::string { return to_string_impl(""); }
[[nodiscard]] auto Color::to_code() const& -> std::string { return to_string_impl("Color::"); }
[[nodiscard]] auto Color::from_string(cpp2::impl::in<std::string_view> s) -> Color{

auto x {s}; 
if ("Black" == x) {return Black; }
else {if ("White" == cpp2::move(x)) {return White; }
#line 1 "../src/color.h2"
}
CPP2_UFCS(report_violation)(cpp2::type_safety, CPP2_UFCS(c_str)(("can't convert string '" + cpp2::to_string(s) + "' to enum of type Color")));
return Black; 
}

[[nodiscard]] auto Color::from_code(cpp2::impl::in<std::string_view> s) -> Color{
std::string str {s}; return from_string(cpp2::string_util::replace_all(cpp2::move(str), "Color::", "")); }
#endif
```

**Comment by gregmarr** [comment]
Is it only when you create the separate header?  There are no issues compiling that here: https://cpp2.godbolt.org/z/81jTWGM63

**Comment by JohelEGP** [comment]
It might be this:
https://github.com/hsutter/cppfront/blob/ccf7011a4ef50eadb1266c5604bbd8d9ad0d87ea/include/cpp2util.h#L371-L376

**Comment by gregmarr** [comment]
@JohelEGP Does `clang` ever set `_MSC_VER`?

**Comment by gregmarr** [comment]
I compared `color.h` above to the output of my cpp2.godbolt.org link, and found that the only differences are the include guard, the `#define CPP2_IMPORT_STD          Yes` and the `#line` directives in `color.h`, and the `main` function in the godbolt output, so no clues there.

**Comment by MatthieuHernandez** [comment]
@JohelEGP and @gregmarr thank you for helping me.
I compile with Clang but the STD comes from MSVC.
The `CPP2_CONSTEXPR` instruction is correctly replaced, however it's the `const` that doesn't work and the `constexpr` that does contrary to what the commentary says.

This doesn't work:
```cpp
inline CPP2_CONSTEXPR Color Color::Black{ 0 }; // error: constexpr constructor never produces a constant expression
```
Because it become this:
```cpp
inline const Color Color::Black{0}; // same error
```
But this work perfectly:
```cpp
inline const Color Color::Black{0}; // same error
```

**Comment by gregmarr** [comment]
I took the command line shown in the original post, removed the `clang  -IC:/Programming/Git/gopp2/../cppfront/include` and `-MT CMakeFiles/gopp2.dir/generate/main.cpp.obj -MF CMakeFiles\gopp2.dir\generate\main.cpp.obj.d -o CMakeFiles/gopp2.dir/generate/main.cpp.obj -c C:/Programming/Git/gopp2/generate/main.cpp` and put it in godbolt.org using `x86-64 clang 16.0.0` with the `color.h` provided and replaced the include of `cpp2util.h` with its current contents, and it still builds fine.

**Comment by gregmarr** [comment]
@MatthieuHernandez I assume your last code block should have `constexpr` instead of `const`.  Yes, just `const` isn't working, but the question is how are you getting that `const`?  You should only get that if `_MSC_VER` is set.

**Comment by MatthieuHernandez** [comment]
I'm on Windows and my standard library comes from MSVC even though I compile with Clang, I think that's where it comes from..

**Comment by MatthieuHernandez** [comment]
Perhaps a Clang under Windows CI pipeline should be added? I saw that there wasn't any.

**Comment by gregmarr** [comment]
@hsutter @MatthieuHernandez As Matthieu mentioned in the PR, there are still other places in that header that assume `_MSC_VER` means MSVC and uses that to trigger workarounds for MSVC issues, several of which would not cause compiler errors and would just result in less performance or capability when using clang on Windows.  I count 22 times that it is used, for 15 distinct locations after eliminating multiple uses on the same line, comments about those uses, and counting a push/pop pair as a single location.

I thought it might be good to have a bit up at the very top of the file that detects the compiler.  Then I found https://github.com/hsutter/cppfront/blob/main/include/cpp2util.h#L335-L368 which I think would do nicely, but I noticed that these are global variables, with no namespace protection, and there probably should be something there.  There should also be something that can be used in conditional compilation.

Maybe we can wrap all of that in a namespace, or at least put something like a `cpp2_compiler_` prefix on the front.  If this is moved to just before the standard library include, we can use something defined there instead of `_MSC_VER`.

However, the compiler alone is not enough to detect that certain headers should not be included, as Matthieu reported that he is building with clang and the MSVC library, so we probably need to detect which library we are using as well.
